src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/SystemService.java

```diff
@@ -184,6 +190,82 @@ public AnchorFile getAnchorFile() throws AnchorNotFoundException {
         return anchorFile;
     }
 
+    /**
+     * Calculate the hex hash of the given anchor file. Used to verify/preview an anchor file before
+     * uploading it
+     * @param anchorBytes
+     * @return
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     */
+    public AnchorFile getAnchorFileFromBytes(byte[] anchorBytes) throws InvalidAnchorInstanceException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        AnchorFile anchorFile = new AnchorFile(calculateAnchorHexHash(anchorBytes));
+        anchorFile.setCreatedAt(FormatUtils.fromDateToOffsetDateTime(anchor.getGeneratedAt()));
+        return anchorFile;
+    }
+
+    /**
+     * Upload a new configuration anchor. A temporary anchor file is created on the filesystem in order to run
+     * the verification process with configuration-client module (via external script).
+     * @param anchorBytes
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     * @throws AnchorUploadException in case of external process exceptions
+     */
+    public void uploadAnchor(byte[] anchorBytes) throws InvalidAnchorInstanceException, AnchorUploadException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        File tempAnchor = null;
+        try {
+            tempAnchor = createTemporaryAnchorFile(anchorBytes);
+            configurationVerifier.verifyInternalConfiguration(tempAnchor.getAbsolutePath());
+            anchorRepository.saveAndReplace(tempAnchor);
+        } catch (ConfigurationVerifier.ConfigurationVerificationException ce) {
+            throw new DeviationAwareRuntimeException(ce, ce.getErrorDeviation());
+        } catch (CodedException ce) {
+            throw ce;
+        } catch (ServiceException | InterruptedException e) {
+            throw new AnchorUploadException(e);
+        } catch (Exception e) {
```

Comment:
Could be 

`} catch (IOException e) {`

and I think more specific one would be better.

Reply:
Don't know but somehow I seem to add `Exception` quite often instead of `IOException`! I updated some javadocs so it's easier to track where the `IOException` is thrown from. In this case it is thrown if the temp file creation fails or saving the anchor file fails. I'm not sure if those require their own exception classes? I probably would not add them.

Reply:
Agreed.